### PR TITLE
Add auto discovery of parameter files when scanning a directory

### PR DESCRIPTION
**Pull request by pim-simons**
## Description
This PR adds auto discovery of parameter files when using the `analyze-directory` command. The code checks if a parameters file with naming according to [Microsofts naming convention](https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/parameter-files#file-name) is present and if so uses this file as the input for the parameters file. 

Added a test `AnalyzeDirectory_ValidInputValues_AnalyzesExpectedNumberOfFilesWithParameters` to check if the parameters file is indeed used. 

I figured I'd try to contribute and help out since this wasn't high in the priority list, if anything is missing or not up to standards coding wise let me know! 

Closes https://github.com/Azure/template-analyzer/issues/296.

---

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [x] The pull request does not introduce breaking changes.

### [General Guidelines](../CONTRIBUTING.md)
- [ ] Title of the pull request is clear and informative.
- [ ] Description of the pull request is clear and informative.
- [x] I have added myself to the 'assignees'.
- [x] I have added 'linked issues' if relevant.

### [Testing Guidelines](../CONTRIBUTING.md#tests)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Comment by pim-simons** [comment]
Regarding `I have added myself to the 'assignees'.`
I can't add myself as an assignee, probably rights related 😉

**Comment by pim-simons** [comment]
@microsoft-github-policy-service agree

**Comment by VeraBE** [comment]
I'm thinking we probably want to do the same for analyze-template, would you mind? Abstracting away the logic of the file search in a function

**Comment by pim-simons** [comment]
> I'm thinking we probably want to do the same for analyze-template, would you mind adding that? Abstracting away the logic of the file search in a function

I added the same functionality in the analyze-template as well, see [`34c1492` (#297)](https://github.com/Azure/template-analyzer/pull/297/commits/34c1492ccd268628ee5600beeea8e11325352dfd)

**Review by VeraBE — APPROVED** [review]

File: src/Analyzer.Cli.FunctionalTests/Tests/ToTestSeparateParametersFile/TemplateWithSeparateParametersFile.bicep [hunk 1]
```diff
@@ -0,0 +1,7 @@
+﻿param infra object
```
Comment (anaismiller) [inline src/Analyzer.Cli.FunctionalTests/Tests/ToTestSeparateParametersFile/TemplateWithSeparateParametersFile.bicep:1]: Would you be able to write the equivalent IaC file in ARM and add it here? I can help if you would like :)

File: src/Analyzer.Cli/CommandLineParser.cs [hunk 2]
```diff
@@ -353,6 +359,18 @@ private IEnumerable<FileInfo> FindTemplateFilesInDirectory(DirectoryInfo directo
             return armTemplates.Concat(bicepTemplates);
         }
 
+        private FileInfo FindParameterFileForTemplate(FileInfo template)
+        {
+            FileInfo parametersFile = null;
+            string defaultParametersFileLocation = Path.Combine(template.Directory.FullName, Path.GetFileNameWithoutExtension(template.Name) + ".parameters.json");
```
Comment (anaismiller) [inline src/Analyzer.Cli/CommandLineParser.cs:365]: [Here](https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/parameter-files#file-name), its mentions that it can also be `azuredeploy.parameters-dev.json` with the `-dev` to identify the development environment. 
Thoughts on making it also work for this case? We can also add this as an issue for a future improvement.
[Method](https://learn.microsoft.com/en-us/dotnet/api/system.io.directory.getfiles?view=net-7.0#system-io-directory-getfiles(system-string-system-string)) and [example](https://stackoverflow.com/questions/1198721/directoryinfo-getfiles-beginning-with)

**Comment by anaismiller** [comment]
Would you also be able to add to our documentation that it uses this pararmeters path? Thank you for adding this in @pim-simons !

File: src/Analyzer.Cli/CommandLineParser.cs (hunk 2 above)
Comment (pim-simons) [reply src/Analyzer.Cli/CommandLineParser.cs]: Good one, hadn't thought of that 👍🏻 
Should now work, see [`b348f20` (#297)](https://github.com/Azure/template-analyzer/pull/297/commits/b348f205123045cb063639db6eed379ad5502301)

**Comment by pim-simons** [comment]
> Would you also be able to add to our documentation that it uses this pararmeters path? Thank you for adding this in @pim-simons !

I updated the documentation, see [`49eff74` (#297)](https://github.com/Azure/template-analyzer/pull/297/commits/49eff744f7de933e879528c6e9c096fde5e8edee)

File: src/Analyzer.Cli.FunctionalTests/Tests/ToTestSeparateParametersFile/TemplateWithSeparateParametersFile.bicep (hunk 1 above)
Comment (pim-simons) [reply src/Analyzer.Cli.FunctionalTests/Tests/ToTestSeparateParametersFile/TemplateWithSeparateParametersFile.bicep:1]: Not really sure if I understand correctly, do you want the Bicep file in ARM in the same location?

File: src/Analyzer.Cli/CommandLineParser.cs [hunk 3]
```diff
@@ -198,10 +198,56 @@ private void SetupCommonOptionsForCommands(List<Command> commands)
...
+            {
+                // Check if a parameters.json file is present according to naming standards here https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/parameter-files#file-name, and if so use it as the parametersFile input
+                var parametersFiles = FindParameterFileForTemplate(templateFilePath);
+                int numOfFilesAnalyzed = 0;
+                bool issueReported = false;
+                var filesFailed = new List<FileInfo>();
+
+                if (parametersFiles.Count() > 0)
+                {
+                    foreach (FileInfo parametersFile in parametersFiles)
```
Comment (VeraBE) [inline src/Analyzer.Cli/CommandLineParser.cs:216]: I see that most of the code inside this `if` was copied from the AnalyzeDirectory function, could you abstract both blocks into a function to avoid code duplication?

File: src/Analyzer.Cli/CommandLineParser.cs [hunk 4]
```diff
@@ -236,16 +282,38 @@ private void SetupCommonOptionsForCommands(List<Command> commands)
...
+                        }
+                    }
                 }
-                else if (res == ExitCode.ErrorAnalysis || res == ExitCode.ErrorInvalidBicepTemplate)
+                else
                 {
-                    filesFailed.Add(file);
+                    ExitCode res = AnalyzeTemplate(file, null);
+
+                    if (res == ExitCode.Success || res == ExitCode.Violation)
```
Comment (VeraBE) [inline src/Analyzer.Cli/CommandLineParser.cs:308]: Also this block is repeated three times in the file

File: src/Analyzer.Cli/CommandLineParser.cs [hunk 5]
```diff
@@ -198,10 +198,56 @@ private void SetupCommonOptionsForCommands(List<Command> commands)
...
+                            numOfFilesAnalyzed++;
+                            issueReported |= res == ExitCode.Violation;
+                        }
+                        else if (res == ExitCode.ErrorAnalysis || res == ExitCode.ErrorInvalidBicepTemplate)
+                        {
+                            filesFailed.Add(templateFilePath);
+                        }
+                    }
+
+                    Console.WriteLine(Environment.NewLine + $"Analyzed {numOfFilesAnalyzed} {(numOfFilesAnalyzed == 1 ? "file" : "files")}.");
```
Comment (VeraBE) [inline src/Analyzer.Cli/CommandLineParser.cs:231]:
```suggestion
                    Console.WriteLine(Environment.NewLine + $"Analyzed the template with {numOfFilesAnalyzed} parameter {(numOfFilesAnalyzed == 1 ? "file" : "files")}.");
```

Or something like that, that expresses that it's the same template combined with different parameter files

File: src/Analyzer.Cli/CommandLineParser.cs (hunk 3 above)
Comment (VeraBE) [inline src/Analyzer.Cli/CommandLineParser.cs:216]: AnalyzeDirectory calls AnalyzeTemplate, and the parameter search is done in both, so if the tool is called with analyze-directory then for each parameter-template combination we would look again at all the combinations

File: src/Analyzer.Cli/CommandLineParser.cs [hunk 6]
```diff
@@ -198,10 +198,56 @@ private void SetupCommonOptionsForCommands(List<Command> commands)
...
+
+                    if (filesFailed.Count > 0)
+                    {
+                        logger.LogError($"Unable to analyze {filesFailed.Count} {(filesFailed.Count == 1 ? "file" : "files")}: {string.Join(", ", filesFailed)}");
```
Comment (VeraBE) [inline src/Analyzer.Cli/CommandLineParser.cs:235]: If a template fails for at least one parameter file, that template will already be reported in the error log from AnalyzeDirectory. We could log something like "Unable to analyze {templateFilePath} with {paramsFailed} parameter file(s): {list of params file paths}", but the information will also be kind of repeated between both error logs, @anaismiller what do you think?

File: src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs [hunk 7]
```diff
@@ -110,7 +126,26 @@ public void AnalyzeDirectory_ValidInputValues_AnalyzesExpectedNumberOfFiles()
...
+            var args = new string[] { "analyze-directory", Path.Combine(Directory.GetCurrentDirectory(), "Tests", "ToTestSeparateParametersFile") };
+
+            using StringWriter outputWriter = new();
+            Console.SetOut(outputWriter);
+
+            var result = _commandLineParser.InvokeCommandLineAPIAsync(args);
+
+            Assert.AreEqual((int)ExitCode.Success, result.Result);
+            StringAssert.Contains(outputWriter.ToString(), "Analyzed 5 files");
+            StringAssert.Contains(outputWriter.ToString(), "Parameters File: " + Path.Combine(Directory.GetCurrentDirectory(), "Tests", "ToTestSeparateParametersFile", "TemplateWithSeparateParametersFile.parameters.json"));
```
Comment (anaismiller) [inline src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs:144]: I'd personally just keep the "parameters.json" and "parameters-dev.json" files as they all have the same values - it feels a little excessive. However, it is nice to know that it worked for all these endings :)

File: src/Analyzer.Cli.FunctionalTests/Tests/ToTestSeparateParametersFile/TemplateWithSeparateParametersFile.bicep (hunk 1 above)
Comment (anaismiller) [reply src/Analyzer.Cli.FunctionalTests/Tests/ToTestSeparateParametersFile/TemplateWithSeparateParametersFile.bicep:1]: Sorry for not being clear. What I meant was:
1. Create a file called "...nctionalTests/Tests/ToTestSeparateParametersFile/TemplateWithSeparateParametersFile.json"

2. Inside the file, write:
{
  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {
    "infra": {
      "type": "object",
      "metadata": {
        "description": "Location for the resource."
      }
    }
  },
  "resources":[
    {
      "type": "Microsoft.Insights/components",
      "apiVersion": "2020-02-02",
      "name": "test",
      "location": "[parameters('infra').environment.location]",
      "kind": "other",
    }
    ]
}

File: src/Analyzer.Cli/CommandLineParser.cs (hunk 6 above)
Comment (anaismiller) [reply src/Analyzer.Cli/CommandLineParser.cs:235]: Less repetition in the error log is better.

File: src/Analyzer.Cli/CommandLineParser.cs [hunk 8]
```diff
@@ -353,6 +421,19 @@ private IEnumerable<FileInfo> FindTemplateFilesInDirectory(DirectoryInfo directo
             return armTemplates.Concat(bicepTemplates);
         }
 
+        private IEnumerable<FileInfo> FindParameterFileForTemplate(FileInfo template)
+        {
+            var parametersFiles = template.Directory.GetFiles(
+                Path.GetFileNameWithoutExtension(template.Name) + ".parameters*.json",
```
Comment (anaismiller) [inline src/Analyzer.Cli/CommandLineParser.cs:427]:
```suggestion
                Path.GetFileNameWithoutExtension(template.Name ".parameters") + "*.json",
```
Not sure if this is better or the way you wrote it is better

File: src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs [hunk 9]
```diff
@@ -110,7 +126,26 @@ public void AnalyzeDirectory_ValidInputValues_AnalyzesExpectedNumberOfFiles()
...
+        {
+            var args = new string[] { "analyze-directory", Path.Combine(Directory.GetCurrentDirectory(), "Tests", "ToTestSeparateParametersFile") };
+
+            using StringWriter outputWriter = new();
+            Console.SetOut(outputWriter);
+
+            var result = _commandLineParser.InvokeCommandLineAPIAsync(args);
+
+            Assert.AreEqual((int)ExitCode.Success, result.Result);
+            StringAssert.Contains(outputWriter.ToString(), "Analyzed 5 files");
```
Comment (VeraBE) [inline src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs:143]: Can you also validate that the template is indeed analyzed only 5 times? For example, counting how many times the template file path shows in the output

File: src/Analyzer.Cli.FunctionalTests/Tests/ToTestSeparateParametersFile/TemplateWithSeparateParametersFile.bicep (hunk 1 above)
Comment (pim-simons) [reply src/Analyzer.Cli.FunctionalTests/Tests/ToTestSeparateParametersFile/TemplateWithSeparateParametersFile.bicep:1]: Ah I see, added in [TemplateWithSeparateParametersFile](https://github.com/Azure/template-analyzer/pull/297/commits/9967bff6ea59906053eff3b4dba913abf1ce9768)

File: src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs (hunk 7 above)
Comment (pim-simons) [reply src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs:144]: A bit excessive indeed, removed in [`9967bff` (#297)](https://github.com/Azure/template-analyzer/pull/297/commits/9967bff6ea59906053eff3b4dba913abf1ce9768)

File: src/Analyzer.Cli/CommandLineParser.cs (hunk 8 above)
Comment (pim-simons) [reply src/Analyzer.Cli/CommandLineParser.cs:427]: Tried to use it your way but that doesn't build, so for now I'll keep it the way it is 😉

File: src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs (hunk 9 above)
Comment (pim-simons) [reply src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs]: Added assertions for this, see [`881b3ad` (#297)](https://github.com/Azure/template-analyzer/pull/297/commits/881b3ada7810321ca1ab57306b974464a531c546)

File: src/Analyzer.Cli/CommandLineParser.cs (hunk 3 above)
Comment (pim-simons) [reply src/Analyzer.Cli/CommandLineParser.cs:216]: Not sure if this is correct, but maybe I am missing something.
From looking at the code we have these relevant methods here:
- `AnalyzeDirectoryCommandHandler`
- `AnalyzeTemplateCommandHandler`
- `AnalyzeTemplate`
- `FindParameterFileForTemplate`

`AnalyzeDirectoryCommandHandler` and `AnalyzeTemplateCommandHandler` both call `AnalyzeTemplate`.
`AnalyzeDirectoryCommandHandler` and `AnalyzeTemplateCommandHandler` both call `FindParameterFileForTemplate`.

`FindParameterFileForTemplate` is not called from `AnalyzeTemplate` as far as I can see. 

If the tool is called with the `analyze-directory` command, the `AnalyzeDirectoryCommandHandler` will search for available parameters files using `FindParameterFileForTemplate`, and then for each combination of template and parameter file call `AnalyzeTemplate`. I don't think we have a situation here where the parameters files are searched for twice. 
Happy to hear of I'm wrong on this though 👍🏻

File: src/Analyzer.Cli/CommandLineParser.cs (hunk 6 above)
Comment (pim-simons) [reply src/Analyzer.Cli/CommandLineParser.cs]: Made a change here [`ff69599` (#297)](https://github.com/Azure/template-analyzer/pull/297/commits/ff69599852c90d50b58726f371e060cb3049666d)

The errors are still logged twice, once initially and the second time during the `Execution summary` logging, but I think that is the whole point of the `Execution summary` logging 🙂 


Logging for `analyze-template` is now:
```
Error: An exception occurred while analyzing template TemplateWithSeparateParametersFile.bicep and parameters file TemplateWithSeparateParametersFile.parameters-dev.json
Error: An exception occurred while analyzing template TemplateWithSeparateParametersFile.bicep and parameters file TemplateWithSeparateParametersFile.parameters-test.json


File: D:\Git\template-analyzer\src\Analyzer.Cli.FunctionalTests\Tests\ToTestSeparateParametersFile\TemplateWithSeparateParametersFile.bicep
Parameters File: D:\Git\template-analyzer\src\Analyzer.Cli.FunctionalTests\Tests\ToTestSeparateParametersFile\TemplateWithSeparateParametersFile.parameters.json

        Rules passed: 0

Analyzed the template with 1 parameter file.

Execution summary:
        The verbose mode (option -v or --verbose) can be used to obtain even more information about the execution.

        Summary of the errors:
                1 instance of: An exception occurred while analyzing template TemplateWithSeparateParametersFile.bicep and parameters file TemplateWithSeparateParametersFile.parameters-dev.json
                1 instance of: An exception occurred while analyzing template TemplateWithSeparateParametersFile.bicep and parameters file TemplateWithSeparateParametersFile.parameters-test.json

        0 Warnings
        2 Errors
```

Logging for `analyze-directory` is now:
```
Directory: D:\Git\template-analyzer\src\Analyzer.Cli.FunctionalTests\Tests\ToTestSeparateParametersFile\
Error: An exception occurred while analyzing template TemplateWithSeparateParametersFile.json and parameters file TemplateWithSeparateParametersFile.parameters-dev.json
Error: An exception occurred while analyzing template TemplateWithSeparateParametersFile.json and parameters file TemplateWithSeparateParametersFile.parameters-test.json


File: D:\Git\template-analyzer\src\Analyzer.Cli.FunctionalTests\Tests\ToTestSeparateParametersFile\TemplateWithSeparateParametersFile.json
Parameters File: D:\Git\template-analyzer\src\Analyzer.Cli.FunctionalTests\Tests\ToTestSeparateParametersFile\TemplateWithSeparateParametersFile.parameters.json

        Rules passed: 0
Error: An exception occurred while analyzing template TemplateWithSeparateParametersFile.bicep and parameters file TemplateWithSeparateParametersFile.parameters-dev.json
Error: An exception occurred while analyzing template TemplateWithSeparateParametersFile.bicep and parameters file TemplateWithSeparateParametersFile.parameters-test.json


File: D:\Git\template-analyzer\src\Analyzer.Cli.FunctionalTests\Tests\ToTestSeparateParametersFile\TemplateWithSeparateParametersFile.bicep
Parameters File: D:\Git\template-analyzer\src\Analyzer.Cli.FunctionalTests\Tests\ToTestSeparateParametersFile\TemplateWithSeparateParametersFile.parameters.json

        Rules passed: 0

Analyzed 2 files.

Execution summary:
        The verbose mode (option -v or --verbose) can be used to obtain even more information about the execution.

        Summary of the errors:
                1 instance of: An exception occurred while analyzing template TemplateWithSeparateParametersFile.json and parameters file TemplateWithSeparateParametersFile.parameters-dev.json
                1 instance of: An exception occurred while analyzing template TemplateWithSeparateParametersFile.json and parameters file TemplateWithSeparateParametersFile.parameters-test.json
                1 instance of: An exception occurred while analyzing template TemplateWithSeparateParametersFile.bicep and parameters file TemplateWithSeparateParametersFile.parameters-dev.json
                1 instance of: An exception occurred while analyzing template TemplateWithSeparateParametersFile.bicep and parameters file TemplateWithSeparateParametersFile.parameters-test.json

        0 Warnings
        4 Errors
```

File: src/Analyzer.Cli/CommandLineParser.cs [hunk 10]
```diff
@@ -282,7 +301,14 @@ private ExitCode AnalyzeTemplate(FileInfo templateFilePath, FileInfo parametersF
             }
             catch (Exception exception)
             {
-                logger.LogError(exception, "An exception occurred while analyzing a template");
+                if (parametersFilePath != null)
+                {
+                    logger.LogError(exception, $"An exception occurred while analyzing template {templateFilePath.Name} and parameters file {parametersFilePath.Name}");
```
Comment (VeraBE) [inline src/Analyzer.Cli/CommandLineParser.cs:306]:
```suggestion
                    logger.LogError(exception, $"An exception occurred while analyzing template {templateFilePath.Name} with parameters file {parametersFilePath.Name}");
```

File: src/Analyzer.Cli/CommandLineParser.cs [hunk 11]
```diff
@@ -198,10 +198,37 @@ private void SetupCommonOptionsForCommands(List<Command> commands)
...
+
+                    exitCode = AnalyzeExitCodes(exitCodes);
+                }
+                else
+                {
+                    exitCode = AnalyzeTemplate(templateFilePath, null);
+                }
+
+                int numOfFilesAnalyzed = exitCodes.Where(x => x == ExitCode.Success || x == ExitCode.Violation).Count();
+                Console.WriteLine(Environment.NewLine + $"Analyzed the template with {numOfFilesAnalyzed} parameter {(numOfFilesAnalyzed == 1 ? "file" : "files")}.");
```
Comment (VeraBE) [inline src/Analyzer.Cli/CommandLineParser.cs:227]: Now that I see the output examples you shared I'm thinking this line is kind of unnecessary, the user will already see that multiple parameters files were used, can we remove it?

File: src/Analyzer.Cli/CommandLineParser.cs [hunk 12]
```diff
@@ -231,36 +258,28 @@ private void SetupCommonOptionsForCommands(List<Command> commands)
...
-                else if (res == ExitCode.ErrorAnalysis || res == ExitCode.ErrorInvalidBicepTemplate)
+                else
                 {
-                    filesFailed.Add(file);
+                    exitCodes.Add(AnalyzeTemplate(file, null));
                 }
             }
 
+            int numOfFilesAnalyzed = exitCodes.Where(x => x == ExitCode.Success || x == ExitCode.Violation).Count();
             Console.WriteLine(Environment.NewLine + $"Analyzed {numOfFilesAnalyzed} {(numOfFilesAnalyzed == 1 ? "file" : "files")}.");
```
Comment (VeraBE) [inline src/Analyzer.Cli/CommandLineParser.cs:280]: I think the idea of this line is for the user to quickly check that nothing was missed, and now we could be missing parameters files too. Maybe it would make sense to say something like "Analyzed x templates and z parameters files"

File: src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs [hunk 13]
```diff
@@ -8,7 +8,7 @@
 using Microsoft.Azure.Templates.Analyzer.Cli;
 using Microsoft.Azure.Templates.Analyzer.Types;
 using Newtonsoft.Json.Linq;
-
+using System.Text.RegularExpressions;
```
Comment (VeraBE) [inline src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs:11]: Can you put this under System.IO? We add references in alphabetical order, except for System ones that go on top

File: src/Analyzer.Cli/CommandLineParser.cs [hunk 14]
```diff
@@ -408,7 +413,7 @@ private IEnumerable<FileInfo> FindTemplateFilesInDirectory(DirectoryInfo directo
                     MatchCasing = MatchCasing.CaseInsensitive,
                     RecurseSubdirectories = true
                 }
-            ).Where(IsValidTemplate);
+            ).Where(s => !s.Name.Contains(".parameters")).Where(IsValidTemplate);
```
Comment (VeraBE) [inline src/Analyzer.Cli/CommandLineParser.cs:369]: Was this change necessary or added just in case? We should have a test that validates AnalyzeDirectory works properly when there are parameters files in the folder

File: src/Analyzer.Cli.FunctionalTests/Analyzer.Cli.FunctionalTests.csproj [hunk 15]
```diff
@@ -28,6 +28,9 @@
     <None Update="Tests\Invalid.bicep">
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
+    <None Update="Tests\ToTestSeparateParametersFile\TemplateWithSeparateParametersFile.json">
```
Comment (VeraBE) [inline src/Analyzer.Cli.FunctionalTests/Analyzer.Cli.FunctionalTests.csproj:31]: The whole Tests folder is copied on line 22, can we revert this change?

File: src/Analyzer.Cli/CommandLineParser.cs (hunk 14 above)
Comment (pim-simons) [reply src/Analyzer.Cli/CommandLineParser.cs:369]: While testing I noticed that the `FindTemplateFilesInDirectory` method would also pick up the parameters files which could then cause issues if these are invalid. Seemed logical to me to try and keep the `FindTemplateFilesInDirectory` method to only try and find the template files, not the parameters, but easily reverted in you think its not necessary 😉

File: src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs (hunk 13 above)
Comment (pim-simons) [reply src/Analyzer.Cli.FunctionalTests/CommandLineParserTests.cs]: done in [`5718325` (#297)](https://github.com/Azure/template-analyzer/pull/297/commits/5718325559dc7b595aa050102a66b33684e1dc30)

File: src/Analyzer.Cli/CommandLineParser.cs [hunk 16]
```diff
@@ -198,10 +198,37 @@ private void SetupCommonOptionsForCommands(List<Command> commands)
...
 
-            var analysisResult = AnalyzeTemplate(templateFilePath, parametersFilePath);
+            ExitCode exitCode;
+            if (parametersFilePath != null)
+            {
+                exitCode = AnalyzeTemplate(templateFilePath, parametersFilePath);
+            }
+            else
+            {
+                // Check if a parameters.json file is present according to naming standards here https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/parameter-files#file-name, and if so use it as the parametersFile input
```
Comment (VeraBE) [inline src/Analyzer.Cli/CommandLineParser.cs:208]: Can we move this comment above FindParameterFileForTemplate instead of having it in two places? Replacing "a parameters.json file is" for "parameters*.json files are"

File: src/Analyzer.Cli/CommandLineParser.cs [hunk 17]
```diff
@@ -198,10 +198,37 @@ private void SetupCommonOptionsForCommands(List<Command> commands)
...
+                var parametersFiles = FindParameterFileForTemplate(templateFilePath);
+                List<ExitCode> exitCodes = new List<ExitCode>();
```
Comment (VeraBE) [inline src/Analyzer.Cli/CommandLineParser.cs:210]:
```suggestion
                var exitCodes = new List<ExitCode>();
```

File: src/Analyzer.Cli.FunctionalTests/Analyzer.Cli.FunctionalTests.csproj (hunk 15 above)
Comment (pim-simons) [reply src/Analyzer.Cli.FunctionalTests/Analyzer.Cli.FunctionalTests.csproj]: Weird, I don't remember setting that config. Anyway, fixed in [`d3dd96d` (#297)](https://github.com/Azure/template-analyzer/pull/297/commits/d3dd96d32e7b79c8c7bd89fe54d9909a95c53700) 😄

File: src/Analyzer.Cli/CommandLineParser.cs [hunk 18]
```diff
@@ -231,36 +258,28 @@ private void SetupCommonOptionsForCommands(List<Command> commands)
             // Log root directory info to be analyzed
             Console.WriteLine(Environment.NewLine + Environment.NewLine + $"Directory: {directoryPath}");
 
-            int numOfFilesAnalyzed = 0;
-            bool issueReported = false;
-            var filesFailed = new List<FileInfo>();
+            List<ExitCode> exitCodes = new List<ExitCode>();
```
Comment (VeraBE) [inline src/Analyzer.Cli/CommandLineParser.cs:261]:
```suggestion
            var exitCodes = new List<ExitCode>();
```

File: src/Analyzer.Cli/CommandLineParser.cs (hunk 11 above)
Comment (pim-simons) [reply src/Analyzer.Cli/CommandLineParser.cs]: Sure, removed in [`12a325b` (#297)](https://github.com/Azure/template-analyzer/pull/297/commits/12a325b393d6dfd29a1a424a299b00206d4d1d79)

File: src/Analyzer.Cli/CommandLineParser.cs (hunk 14 above)
Comment (VeraBE) [reply src/Analyzer.Cli/CommandLineParser.cs:368]: Oh D: Yes, makes sense, we can add a test for that later, we should have done it before

File: src/Analyzer.Cli/CommandLineParser.cs (hunk 11 above)
Comment (VeraBE) [reply src/Analyzer.Cli/CommandLineParser.cs]: Also the previous line, right?

Comment (pim-simons) [reply src/Analyzer.Cli/CommandLineParser.cs]: Oops you're right, my bad, it's getting late (at least in my timezone 😉), fixed in [`e040c73` (#297)](https://github.com/Azure/template-analyzer/pull/297/commits/e040c73f12817287c5b07aa77d96fc912e180111)

File: src/Analyzer.Cli/CommandLineParser.cs [hunk 19]
```diff
@@ -282,7 +296,14 @@ private ExitCode AnalyzeTemplate(FileInfo templateFilePath, FileInfo parametersF
             }
             catch (Exception exception)
             {
-                logger.LogError(exception, "An exception occurred while analyzing a template");
+                if (parametersFilePath != null)
+                {
+                    logger.LogError(exception, $"An exception occurred while analyzing template {templateFilePath.Name} with parameters file {parametersFilePath.Name}");
```
Comment (VeraBE) [inline src/Analyzer.Cli/CommandLineParser.cs:301]: Oh, sorry, I just realized that the message could be unclear if there are multiple folders, can you use FullName instead of Name in both variables here and in line 305?

File: src/Analyzer.Cli/CommandLineParser.cs [hunk 20]
```diff
@@ -231,36 +254,27 @@ private void SetupCommonOptionsForCommands(List<Command> commands)
...
-            if (filesFailed.Count > 0)
-            {
-                logger.LogError($"Unable to analyze {filesFailed.Count} {(filesFailed.Count == 1 ? "file" : "files")}: {string.Join(", ", filesFailed)}");
-                exitCode = issueReported ? ExitCode.ErrorAndViolation : ExitCode.ErrorAnalysis;
-            }
-            else
-            {
-                exitCode = issueReported ? ExitCode.Violation : ExitCode.Success;
-            }
+            ExitCode exitCode = AnalyzeExitCodes(exitCodes);
```
Comment (anaismiller) [inline src/Analyzer.Cli/CommandLineParser.cs:277]:
```suggestion
            var exitCode = AnalyzeExitCodes(exitCodes);
```

**Review by anaismiller — APPROVED** [review]
Minor changes needed for me, approving!

File: src/Analyzer.Cli/CommandLineParser.cs (hunk 19 above)
Comment (pim-simons) [reply src/Analyzer.Cli/CommandLineParser.cs]: Changed in [`c62c0d7` (#297)](https://github.com/Azure/template-analyzer/pull/297/commits/c62c0d7cd7fcfd0904a120161940a49bdff76322)

**Review by VeraBE — APPROVED** [review]
